cpp/autosar/src/rules/A7-3-1/HiddenInheritedNonOverridableMemberFunction.ql

```diff
@@ -15,24 +15,43 @@
 
 import cpp
 import codingstandards.cpp.autosar
+import codingstandards.cpp.Class
 
-from FunctionDeclarationEntry overridingDecl, FunctionDeclarationEntry hiddenDecl
+/**
+ * Holds if the class has a non-virtual member function with the given name.
+ */
+pragma[noinline, nomagic]
+predicate hasNonVirtualMemberFunction(Class clazz, MemberFunction mf, string name) {
+  mf.getDeclaringType() = clazz and
+  mf.getName() = name and
+  not mf.isVirtual()
+}
+
+/**
+ * Holds if the member function is in a class with the given base class, and has the given name.
+ */
+pragma[noinline, nomagic]
+predicate hasDeclarationBaseClass(MemberFunction mf, Class baseClass, string functionName) {
+  baseClass = mf.getDeclaringType().getABaseClass() and
+  functionName = mf.getName()
+}
+
+from MemberFunction overridingDecl, MemberFunction hiddenDecl, Class baseClass, string name
 where
   not isExcluded(overridingDecl, ScopePackage::hiddenInheritedNonOverridableMemberFunctionQuery()) and
   // Check if we are overriding a non-virtual inherited member function
-  overridingDecl.getName() = hiddenDecl.getName() and
-  overridingDecl.getDeclaration().getDeclaringType().getABaseClass() =
-    hiddenDecl.getDeclaration().getDeclaringType() and
-  not hiddenDecl.getDeclaration().isVirtual() and
+  hasNonVirtualMemberFunction(baseClass, hiddenDecl, name) and
+  hasDeclarationBaseClass(overridingDecl, baseClass, name) and
   // Where the hidden member function isn't explicitly brought in scope through a using declaration.
   not exists(UsingDeclarationEntry ude |
-    ude.getDeclaration() = hiddenDecl.getDeclaration() and
-    ude.getEnclosingElement() = overridingDecl.getDeclaration().getDeclaringType() and
-    ude.getLocation().getStartLine() < overridingDecl.getLocation().getStartLine()
+    ude.getDeclaration() = hiddenDecl and
+    ude.getEnclosingElement() = overridingDecl.getDeclaringType()
   ) and
   // Exclude compiler generated member functions which include things like copy constructor that hide base class
   // copy constructors.
-  not overridingDecl.getDeclaration().isCompilerGenerated()
+  not overridingDecl.isCompilerGenerated() and
```

Comment:
The OpenPilot results show we are finding issues in macros like `QObject`.
The suggestion removes 38 of those results.

```suggestion
  not overridingDecl.isCompilerGenerated() and
  not overridingDecl.isInMacroExpansion() and
```

Reply:
I don't think we should generally exclude results for this rule because they were generated by a macro. If I use a macro to generate a member function that hides base class member functions it still contravenes the rule, and addressable by adding a "using ...".

For OpenPilot, QT has a bit of an unusual meta-object system, implemented through the `Q_OBJECT` macro. This macro generates common member functions for QT objects. Reviewing the results for this rule we seemed to be flagging 3 generated functions: `qt_static_metacall`, `tr` and `trUtf8`. `tr` and `trUtf8` seem to be genuine contraventions of this rule - as the macro is generating an implementation of a non-virtual member function that hides the base class implementation.

`qt_static_metacall` is interesting because it's actually a private (static) member function. I don't think we should be reporting _private_ member functions from the base class, because they would not be visible in the subclass anyway, so no hiding is going on. I've therefore updated the query to exclude this case.